### PR TITLE
Music Profiles Example

### PR DESCRIPTION
**Pull request by vasia**
Hi everyone!

This is an initial implementation for the example of #23.
It also contains an additional example -Label Propagation- which is used as the community detection algorithm for the last step.
For the sake of the example I have used a made-up toy dataset, but I'm planning to test this with the echo nest dataset on a cluster.

Comments are welcome!
-V.

### CODE REVIEW

**Comment by vasia** [comment]
@ktzoumas test

File: src/main/java/flink/graphs/example/LabelPropagationExample.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    private static long numVertices = 20;
+    private static int maxIterations = 10;
+    private static int numberOfLabels = 3;
+
+    @SuppressWarnings("serial")
+	private static DataSet<Vertex<Long, Long>> getVertexDataSet(ExecutionEnvironment env) {
+            return env.generateSequence(1, numVertices)
+                    .map(new MapFunction<Long, Vertex<Long, Long>>() {
+                        public Vertex<Long, Long> map(Long l) throws Exception {
+                        	Random randomGenerator = new Random();
```
Comment (StephanEwen) [inline src/main/java/flink/graphs/example/LabelPropagationExample.java]: I think random number generator are not totally cheap to initialize, so it is better to define them outside the repeatedly called function.

**Comment by StephanEwen** [comment]
Concerning the LabelPropagation algorithm: This one does not exploit the delta iteration behavior, because it always sets a new label, without checking if it is any different than the previous one.

Also: Is it guaranteed to converge? Or can there be a corner case where the most frequent neighbor label oscillates between two groups of vertices.

When you change it to exploit delta iteration behavior: This may have tricky implications on the "select most frequent neighbor label" strategy, as you may not see every neighbor in later steps (when they converged and do not send their label anymore).

File: src/main/java/flink/graphs/example/MusicProfiles.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    	 */
+    	DataSet<Vertex<String, Long>> verticesWithCommunity = similarUsersGraph.mapVertices(
+    			new InitVertexLabels(numberOfLabels))
+    			.run(new LabelPropagation<String>(numIterations)).getVertices();
+    	verticesWithCommunity.print();
+
+    	env.execute();
+    }
+
+    @SuppressWarnings("serial")
```
Comment (StephanEwen) [inline src/main/java/flink/graphs/example/MusicProfiles.java]: You can have a single `@SuppressWarnings("serial")` above the surrounding class. Just cosmetics...

File: src/main/java/flink/graphs/example/MusicProfiles.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	public static final class ExtractMismatchSongIds implements MapFunction<String, Tuple1<String>> {
+		public Tuple1<String> map(String value) {
+			String[] tokens = value.split("\\s+"); 
+			String songId = tokens[1].substring(1);
+			return new Tuple1<String>(songId);
+		}
+    }
+
+    @SuppressWarnings("serial")
+	public static final class FilterOutMismatches implements CoGroupFunction<Tuple3<String, String, Integer>,
```
Comment (StephanEwen) [inline src/main/java/flink/graphs/example/MusicProfiles.java]: This is actually an *anti-join* like in a *where not exists* subquery. We should add this to the APIs, it seems a common pattern.

**Comment by vasia** [comment]
The label propagation implementation is actually quite simplified :)
The original algorithm ([see paper](http://arxiv.org/pdf/0709.2938.pdf)) is guaranteed to converge, but has 2 important differences: (a) initially every node has a different label and (b) a way to break ties needs to be defined.

Regarding the delta behavior: In the general case, the "most-frequent" operator is not incremental. For using a delta iteration, we should store something like a map of all previously seen labels with frequencies as the vertex value and then increase/decrease these frequencies, when only receiving messages from changed neighbors. The tricky part is that maintaining this state might be more expensive than using a bulk iteration. 
In fact, this was one of the algorithms I had chosen in order to show the need for dependency iterations (only retrieve neighbor values if at least one of them has changed value). In some experiments I had run (maybe 4-5 months ago?) this was better than bulk and delta, but it also depends on the graph structure.

In any case, I would propose the following: I will change the implementation to make it follow the description of the algorithm from the paper to guarantee convergence, but not use the delta version for the moment. I think this is fair enough for an example ;)
Later, we can experiment a bit and design an optimized version for the graph library. What do you think?

File: src/main/java/flink/graphs/example/MusicProfiles.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    	 *  Filter out the mismatches from the triplets dataset
+    	 */
+    	DataSet<Tuple3<String, String, Integer>> validTriplets = triplets.coGroup(mismatches)
+    			.where(1).equalTo(0).with(new FilterOutMismatches());
+
+    	/**
+    	 *  Create a user -> song weighted bipartite graph
+    	 *  where the edge weights correspond to play counts
+    	 */
+    	DataSet<Edge<String, Integer>> userSongEdges = validTriplets.map(new Tuple3ToEdgeMap<String, Integer>());
```
Comment (StephanEwen) [inline src/main/java/flink/graphs/example/MusicProfiles.java]: Would it be good if the graph had a constructor from Tuple3 and do the conversion to the Edge by itself?

Comment (vasia) [inline src/main/java/flink/graphs/example/MusicProfiles.java]: yes, we actually thought the same yesterday with @ktzoumas. In the to-do list :)

**Comment by StephanEwen** [comment]
Regarding the Label Propagation: sounds good

**Comment by StephanEwen** [comment]
This is a very nice start :-) Two things came to my mind while reading this code:

1) All functions are defined somewhere else as inner classes. While that was previously necessary because of the non-serializable closure, the closure cleaning should solve this. The code is easier to read with anonymous inner classes, because you do not have to jump between function call and definition all the time. On the other hand, it introduces a hard requirement that closure cleaning is activated. Users that de-activate it on the environment (for whatever reason) will have problems.

2) Maybe the part about connected users could be extended a bit. Right now, everyone who ever shared a song is connected. Is everybody connected since Gangnam style? Maybe it makes sense to define it a bit different, a number connections above a the mean by some variance or so.

File: src/main/java/flink/graphs/example/MusicProfiles.java (hunk 4 above)
Comment (cebe) [inline src/main/java/flink/graphs/example/MusicProfiles.java]: > Would it be good if the graph had a constructor from Tuple3 and do the conversion to the Edge by itself?

Will come soon: https://github.com/project-flink/flink-graph/issues/53#issuecomment-69756322 ;)

**Comment by vasia** [comment]
Thanks for all the feedback! 

(1) I actually prefer the inner classes approach for the examples. If the names of the classes are descriptive enough and there are comments, I think it makes it much easier to quickly get an idea of what is happening. Also, I think it's easier to show how to use the gelly methods ;)

(2) This part can be extended a lot and I do have plans to do so :) We could have a threshold of how many common songs make two users connected, take into account the playcounts of each song, define a max number of neighbors per user, etc.

**Pull request closed without merging**